poetry/core/packages/package_specification.py

```diff
@@ -0,0 +1,118 @@
+from typing import FrozenSet
+from typing import List
+from typing import Optional
+
+from poetry.core.utils.helpers import canonicalize_name
+
+
+class PackageSpecification(object):
+    def __init__(
+        self,
+        name,
+        source_type=None,
+        source_url=None,
+        source_reference=None,
+        source_resolved_reference=None,
+        features=None,
+    ):  # type: (str, Optional[str], Optional[str], Optional[str], Optional[str], Optional[List[str]]) -> None
+        self._pretty_name = name
+        self._name = canonicalize_name(name)
+        self._source_type = source_type
+        self._source_url = source_url
+        self._source_reference = source_reference
+        self._source_resolved_reference = source_resolved_reference
+
+        if not features:
+            features = []
+
+        self._features = frozenset(features)
+
+    @property
+    def name(self):  # type: () -> str
+        return self._name
+
+    @property
+    def pretty_name(self):  # type: () -> str
+        return self._pretty_name
+
+    @property
+    def complete_name(self):  # type () -> str
+        name = self._name
+
+        if self._features:
+            name = "{}[{}]".format(name, ",".join(sorted(self._features)))
+
+        return name
+
+    @property
+    def source_type(self):  # type: () -> Optional[str]
+        return self._source_type
+
+    @property
+    def source_url(self):  # type: () -> Optional[str]
+        return self._source_url
```

Comment:
At least [here](https://github.com/python-poetry/poetry/blob/17b165bf8c8d989d672417fd4610721df59ae1d0/poetry/repositories/installed_repository.py#L146-L156) we also need setters for these properties. Don't know about the others for now :)

Reply:
Eventually, I want them to be set at objects' initialization and not be settable (see my comment above about sources objects).

I don't want to give the impression that settings them is part of a public API that people can use.

Reply:
Woops wrong button. :(

@sdispater we should really discuss what an "api" is for both core and poetry. Define it, document it and refactor to make it possible at a target release.

Reply:
@abn I think we will be able to discuss this API after the 1.1 release. However, what I can say is that the versioning of Poetry follows semantic versioning only for the CLI at the moment, even though we try to have a sensible backwards compatibility policy.

Reply:
Considering that we will need to change code anyway with the `PackageSpecification` class I think explicitly setting `_source_url` for now should be fine.